# Content List


01 Importing Libraries
02 Importing Dataframes
03 Mixed Data Type
04 Finding and Addressing Missing Values
05 Finding and Addressing Duplicates
06 Exercises 2 - 9 

# 01 Importing Libraries

In [2]:
import pandas as pd
import numpy as np
import os

# 02 Importing Dataframes

In [7]:
path = r'/Users/a/Desktop/4'



In [8]:
path

'/Users/a/Desktop/4'

In [9]:
df_prods = pd.read_csv(os.path.join(path, 'products.csv'), index_col = False)

In [10]:
path = r'/Users/a/Career Foundry/Achievement 4/02 Data/02 Prepared Data'

In [11]:
path

'/Users/a/Career Foundry/Achievement 4/02 Data/02 Prepared Data'

In [12]:
df_ords = pd.read_csv(os.path.join(path, 'orders_wrangled.csv'), index_col = False)

In [13]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   time_orders  days_since_prior_order  
0            8                     NaN  
1            7                    15.0  
2           12                    21.0  
3            7                    29.0  
4           15                    28.0

# 03 Mixed Type Data

In [18]:
# Create a dataframe
df_test = pd.DataFrame() 

In [19]:
# Create a mixed type column
df_test['mix'] = ['a','b', 1, True]

In [20]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [21]:
# Check for mixed data types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [22]:
df_test['mix'] = df_test['mix'].astype('str')


# 04 Finding and Addressing Missing Values

In [23]:
# Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [24]:
df_nan = df_prods[df_prods['product_name'].isnull() == True] 


In [25]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [26]:
df_prods.shape

(49693, 5)

In [27]:
# Addressing missing values
# creating a new dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]


In [28]:
df_prods_clean.shape

(49677, 5)

In [29]:
# Creating a boolean column 

In [33]:
df_prods['ordered_today'] = np.where(df_prods['days_prior_order'].isnull(), True, False)

KeyError: 'days_prior_order'

# 05 Finding and Addressing Duplicates

In [34]:
# Finding Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()] 


In [35]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [36]:
df_prods_clean.shape

(49677, 5)

In [37]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [38]:
df_prods_clean_no_dups.shape

(49672, 5)

# 06 Exercises 2 - 9 

In [39]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The maximum price under the prices column is unusually high for a grocery product

In [40]:
# 3. Check for mixed-type data in your df_ords dataframe.


In [41]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
    


In [42]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   time_orders  days_since_prior_order  
0            8                     NaN  
1            7                    15.0  
2           12                    21.0  
3            7                    29.0  
4           15                    28.0

4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

There are no mixed data types columns

In [44]:
# 5. Run a check for missing values in your df_ords dataframe.
df_ords.isnull().sum()


Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
time_orders                    0
days_since_prior_order    206209
dtype: int64

The column days since prior order has 206209 missing values

In [59]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True] 


In [60]:
df_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         time_orders  days_since_prior_order  
0                  8                     NaN  
11                11                     NaN  
26                14                     NaN  
39                11                     NaN  
45                12                     NaN  
...              ...                     ...  
3420930           12                     NaN  
3420934           18                     NaN  
3421002           19                     NaN  
3421019           15                     NaN  
3421069           11                     NaN  

[206209 rows x 7 columns]

Meaning: the customer has placed an order at one day so there are a lot of missing entries for the days before the order was places

In [ ]:
# 6. Address the missing values using an appropriate method.


We can´t impute the data because it is string data. I don´t find this specific variable relevant for the analysis, I would simply entry the data of when the order has been placed. In that case, I would solve the issue of missing data. Another way is to create a boolean flaf for orders that have been placed today

In [14]:
df_ords['ordered_today']= np.where(df_ords['days_since_prior_order'].isnull(), True, False)

In [15]:
df_ords.shape

(3421083, 8)

# 7. Run a check for duplicate values in your df_ords data.


In [17]:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]




In [18]:
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [19]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, time_orders, days_since_prior_order, ordered_today]
Index: []

In [20]:
df_ords_dups.shape

(0, 8)

There are no duplicates

In [ ]:
# 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.


In [21]:
path = r'/Users/a/Career Foundry/Achievement 4'

In [22]:
path

'/Users/a/Career Foundry/Achievement 4'

In [23]:
df_ords_clean.to_csv(os.path.join(path, '02 Data','02 Prepared Data','orders_cleaned.csv'))

In [78]:
df_prods_clean.to_csv(os.path.join(path, '02 Data','02 Prepared Data','products_cleaned.csv'))